### PR TITLE
when `x` is `NaN` in trig functions return `x` rather than `NaN`

### PR DESCRIPTION
**Pull request by oscardssmith**
see https://github.com/JuliaLang/julia/issues/48523 for reasoning.

### CODE REVIEW

**Comment by topolarity** [comment]
I think this guarantee is destroyed by the LLVM optimizer anyway

From page 5 of https://arxiv.org/pdf/1603.09290.pdf:
> The standard corresponds to IEEE 754-2008 but it only defines a single NaN value and does not distinguish between signalling and quiet NaNs. Thus, our implementation cannot verify whether an operation with NaN operands returns one of the input NaNs, propagating debug information encoded in the NaN, as recommended by the IEEE standard. In practice, LLVM does not attempt to preserve information in NaNs, so this limitation does not affect our ability to verify LLVM optimizations.

IMO that recommendation from the IEEE standard is silly. Although the discussion in #48523 rightly points out that payload _tagging_ has useful applications, I don't think that extends to _propagating_ the payload through arithmetic.

**Comment by oscardssmith** [comment]
yeah. This is (I believe) a minor performance benefit,  but also makes NaN propagation a little more reliable (although LLVM still is allowed to mess with us)

**Comment by Seelengrab** [comment]
The "return `x` on `NaN` input" policy is consistent with x86, which (I think) is what we're doing in `@fastmath`, right? See also https://www.felixcloutier.com/x86/minps, though trigonometric equivalents in particular have.. underspecified behavior, raising floating point exceptions on SNaN, and unspecified behavior on QNaN (see https://www.felixcloutier.com/x86/fcos and https://www.felixcloutier.com/x86/fsincos)...

File: base/special/trig.jl [hunk 1]
```diff
@@ -583,7 +583,7 @@ function atan(y::T, x::T) where T<:Union{Float32, Float64}
     #    S9) ATAN2(+-INF,-INF ) is +-3pi/4;
     #    S10) ATAN2(+-INF, (anything but,0,NaN, and INF)) is +-pi/2;
     if isnan(x) || isnan(y) # S1 or S2
```
Comment (mikmoore) [inline base/special/trig.jl:585]:
```suggestion
    if isnan(x) | isnan(y) # S1 or S2
```
`isnan(x) | isnan(y)` tends to compile nicer than `isnan(x) || isnan(y)`, at least on x86. For me, this saves 1 comparison and 1 branch.

**Pull request merged**